File: framework/languages/dotnet/e2e/YieldDispatch/Client/Scenarios/YdC2TimerReentryScenario.cs
using YieldDispatch.Client.Support;
using YieldDispatch.Shared;

namespace YieldDispatch.Client.Scenarios;

internal static class YdC2TimerReentryScenario
{
    public static async Task<string> RunAsync(IZlinkStreamConnector client, string spotRid)
    {
        var requestId = $"YD-C2-{Guid.NewGuid():N}";
        client.Send(new TimerStartMsg(requestId, $"{requestId}-same", "yield-then-next", 50, 350))
            .PacketName("TimerStartMsg")
            .Metadata(YieldDispatchNames.SpotRidMetadata, spotRid).Submit();
        var evidence = await client.Request(new YieldEvidenceWaitReq(requestId, "timer-next-completed"))
            .PacketName("YieldEvidenceWaitReq")
            .Metadata(YieldDispatchNames.TargetNodeRidMetadata, "play-a")
            .Timeout(TimeSpan.FromSeconds(30))
            .Async<YieldEvidenceRes>();
        ScenarioAssert.ContainsExactRequestInOrder(evidence.Evidence, requestId, [
            "timer-yield-started",
            "timer-yield-released",
            "timer-yield-resumed",
            "timer-yield-completed",
            "timer-next-started",
            "timer-next-completed"
        ]);
        client.Send(new TimerStopMsg(requestId))
            .PacketName("TimerStopMsg")
            .Metadata(YieldDispatchNames.SpotRidMetadata, spotRid).Submit();
        return requestId;
    }
}

File: framework/languages/node/e2e/YieldDispatch/Client/Scenarios/yd-c2-timer-reentry-scenario.ts
import type { TimerStartMsg } from '../../Shared/messages';
import { containsRequestMarkersInOrder } from '../Support/scenario-assert';
import type { ZlinkStreamConnector } from '@zlink-systems/stream-connector';
import { startTimer, stopTimers, waitForEvidence } from './yd-c1-timer-isolation-scenario';

export async function runYdC2(client: ZlinkStreamConnector, spotRid: string): Promise<string> {
  const requestId = `YD-C2-${uniqueId()}`;
  await startTimer(client, spotRid, {
    requestId,
    timerName: `${requestId}-same`,
    mode: 'yield-then-next',
    periodMs: 50,
    delayMs: 350
  } satisfies TimerStartMsg);
  const evidence = await waitForEvidence(client, requestId, 'timer-next-completed');
  containsRequestMarkersInOrder(evidence.evidence, requestId, [
    'timer-yield-started',
    'timer-yield-released',
    'timer-yield-resumed',
    'timer-yield-completed',
    'timer-next-started',
    'timer-next-completed'
  ], 'YD-C2 marker order mismatch.');
  await stopTimers(client, spotRid, requestId);
  console.log('scenario YD-C2 passed');
  return requestId;
}

function uniqueId(): string {
  return `${Date.now().toString(16)}-${Math.random().toString(16).slice(2)}`;
}
